### PR TITLE
add session struct in server proto

### PR DESCRIPTION
**Pull request by Yancey0623**
Related issue #531

### CODE REVIEW

File: server/sqlflowserver.go [hunk 1]
```diff
@@ -46,12 +46,24 @@ type Server struct {
 
 // Run implements `rpc Run (Request) returns (stream Response)`
 func (s *Server) Run(req *pb.Request, stream pb.SQLFlow_RunServer) error {
-	pr := s.run(req.Sql, s.db, s.modelDir)
+	db := s.db
+	var err error
+
+	// TOOD(Yancey1989): delete the IF statment
+	// when we enable session in the pysqlflow repo by default.
+	if req.Session != nil {
```
Comment (typhoonzero) [inline server/sqlflowserver.go]: Maybe add a global settings variable when starting the sqlflowserver process.

File: server/sqlflowserver.go [hunk 2]
```diff
@@ -74,6 +86,23 @@ func (s *Server) Run(req *pb.Request, stream pb.SQLFlow_RunServer) error {
...
 }
 
+func dbConnStrWithCredencial(token, dbConnStr string) string {
+	// TODO(Yancey1989): implement fetch Database crediencial according
+	// to the user token.
+	return dbConnStr
+}
+
+// SessionDB returns a sql.DB object accoding to the request session.
+func sessionDB(session *pb.Session) (*sf.DB, error) {
```
Comment (typhoonzero) [inline server/sqlflowserver.go]: sessionDB => `getCachedDBConn`

and should add a global mapping to store the cached connections.

File: server/sqlflowserver_test.go [hunk 3]
```diff
@@ -143,6 +143,19 @@ func TestGoroutineLeaky(t *testing.T) {
 	}
 }
 
+func TestSessionDB(t *testing.T) {
```
Comment (typhoonzero) [inline server/sqlflowserver_test.go]: maybe also need to test sending session message throught RPC

File: server/sqlflowserver.go (hunk 2 above)
Comment (Yancey0623) [reply server/sqlflowserver.go]: done.

File: server/sqlflowserver.go (hunk 1 above)
Comment (Yancey0623) [reply server/sqlflowserver.go]: Maybe we can enable session when the Flag `--datasource` is an empty string, that we don't need another new flag.

File: sql/database_cache.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import "time"
+
+// DBConnCache caches the DB object
+type DBConnCache struct {
```
Comment (tonyyang-svail) [inline sql/database_cache.go]: FYI, the Hive driver and ODPS driver don't maintain long connections on each `DB` object. So adding cache may not help on the performance.

File: sql/database_cache.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2019 The SQLFlow Authors. All rights reserved.
```
Comment (wangkuiyi) [inline sql/database_cache.go]: The cache is nothing related to "session struct". It should be in a separate PR, one with thorough unit test and a benchmark to verify the performance improvement. Please remove the cache from this PR.

**Review by wangkuiyi — CHANGES_REQUESTED** [review]

File: sql/database_cache.go (hunk 4 above)
Comment (Yancey0623) [reply sql/database_cache.go]: Not only the performance,  to cache the `DB` object can save the SQL context, such as `USE DB1` if we can load balance the requests to the same SQLFlow server instance according to the `Token`. And as the comment: https://github.com/sql-machine-learning/sqlflow/pull/530#discussion_r296898832 ,  I will remove the `db_conn_cache` from this PR, and implement it in another PR with a benchmark.

File: sql/database_cache.go (hunk 5 above)
Comment (Yancey0623) [reply sql/database_cache.go]: Done, removed the `DBCache`.

File: server/sqlflowserver_test.go (hunk 3 above)
Comment (Yancey0623) [reply server/sqlflowserver_test.go]: done.

File: server/proto/sqlflow.proto [hunk 6]
```diff
@@ -19,13 +19,19 @@ service SQLFlow {
     rpc Run (Request) returns (stream Response);
 }
 
+message Session {
```
Comment (wangkuiyi) [inline server/proto/sqlflow.proto:22]: Is it right to name this struct `Session`? Should we rename it into `Cookies`?

According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies

> An HTTP cookie (web cookie, browser cookie) is a small piece of data that a server sends to the user's web browser. The browser may store it and send it back with the next request to the same server. 

It sounds similar to what this struct is meant to.

**Review by wangkuiyi — APPROVED** [review]

File: server/proto/sqlflow.proto (hunk 6 above)
Comment (Yancey0623) [reply server/proto/sqlflow.proto:22]: Seems `session` is more appropriate: 
1. a cookie is the concept of HTTP proto, a session is the Broader concept, according to https://en.wikipedia.org/wiki/Session_(computer_science)
     > a session is a temporary and interactive information interchange between two or more communicating devices, or between a computer and user

1. We also want to store the SQL context on the server side in the session struct in the future,  and the cookie is stored on the browser, according to the design: https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/auth_design.md#session 

1. This PR is a crude implement of Session, hasn't included saving the context on the server side, because of: 
    >  the Hive driver and ODPS driver don't maintain long connections on each DB object

    according to the comment: https://github.com/sql-machine-learning/sqlflow/pull/530#discussion_r296859071 , and I think we need to implement this in the future.

**Review by typhoonzero — APPROVED** [review]
LGTM++
